core/src/apps/webauthn/__init__.py

```diff
@@ -749,3 +1230,413 @@ def msg_error(cid: int, code: int) -> Cmd:
+        user_presence = options.get("up", True)
+        user_verification = options.get("uv", False)
+
+        # Get extensions.
+        hmac_secret = param.get(_GETASSERT_CMD_EXTENSIONS, {}).get("hmac-secret", None)
+
+        client_data_hash = param[_GETASSERT_CMD_CLIENT_DATA_HASH]
+    except Exception:
+        return cbor_error(req.cid, _ERR_INVALID_CBOR)
+
+    # Check data types.
+    if (
+        not isinstance(hmac_secret, (dict, type(None)))
+        or not isinstance(client_data_hash, (bytes, bytearray))
+        or not isinstance(user_presence, bool)
+        or not isinstance(user_verification, bool)
+    ):
+        return cbor_error(req.cid, _ERR_INVALID_CBOR)
+
+    # User verification cannot happen without user presence.
+    user_presence = user_presence or user_verification
+
+    if user_verification and not config.has_pin():
+        # User verification requested, but PIN is not enabled.
+        state_set = dialog_mgr.set_state(Fido2ConfirmNoPin(req.cid, dialog_mgr.iface))
+    elif not cred_list:
+        # No credentials. This authenticator is not registered.
+        if user_presence:
+            state_set = dialog_mgr.set_state(
+                Fido2ConfirmNoCredentials(req.cid, dialog_mgr.iface, rp_id)
+            )
+        else:
+            return cbor_error(req.cid, _ERR_NO_CREDENTIALS)
+    elif not user_presence:
+        # Silent authentication.
+        try:
+            response_data = cbor_get_assertion_sign(
+                client_data_hash, rp_id_hash, cred_list[0], hmac_secret, False, False
+            )
+            return Cmd(req.cid, _CMD_CBOR, bytes([_ERR_NONE]) + response_data)
+        except Exception:
+            return cbor_error(req.cid, _ERR_OPERATION_DENIED)
+    else:
+        # Ask user to confirm one of the credentials.
+        state_set = dialog_mgr.set_state(
+            Fido2ConfirmGetAssertion(
+                req.cid,
+                dialog_mgr.iface,
+                client_data_hash,
+                cred_list,
+                hmac_secret,
+                user_verification,
+            )
+        )
+
+    if not state_set:
+        return cmd_error(req.cid, _ERR_CHANNEL_BUSY)
+
+    return None
+
+
+def cbor_get_assertion_hmac_secret(
+    cred: Credential, hmac_secret: dict
+) -> Optional[bytes]:
+    key_agreement = hmac_secret[1]  # The public key of platform key agreement key.
+    salt_enc = hmac_secret[2]  # The encrypted salt.
+    salt_auth = hmac_secret[3]  # The HMAC of the encrypted salt.
+
+    x = key_agreement[_COSE_X_COORD_KEY]
+    y = key_agreement[_COSE_Y_COORD_KEY]
+    if (
+        key_agreement[_COSE_ALG_KEY] != _COSE_ALG_ES256
+        or key_agreement[_COSE_KEY_TYPE_KEY] != _COSE_KEY_TYPE_EC2
+        or key_agreement[_COSE_CURVE_KEY] != _COSE_CURVE_P256
+        or len(x) != 32
+        or len(y) != 32
+        or len(salt_auth) != 16
+        or len(salt_enc) not in (32, 64)
+    ):
+        raise ValueError
```

Comment:
I'm not sure what is the correct way to handle failure in extension processing. Nevertheless, ignoring such a extension is completely all right:
> All WebAuthn extensions are OPTIONAL for both clients and authenticators. Thus, any extensions requested by a Relying Party MAY be ignored by the client browser or OS and not passed to the authenticator at all, or they MAY be ignored by the authenticator. Ignoring an extension is never considered a failure in WebAuthn API processing, so when Relying Parties include extensions with any API calls, they MUST be prepared to handle cases where some or all of those extensions are ignored.

Reply:
ACK